src/Common/src/System/Net/SafeCloseSocket.Unix.cs

```diff
@@ -208,6 +208,15 @@ private void InnerReleaseHandle()
             }
         }
 
+        private void UnblockSocket(InnerSafeCloseSocket innerSocket)
+        {
+            if ((AsyncContext == null || !AsyncContext.GetNonBlocking()) && innerSocket != null && !_underlyingHandleNonBlocking && !innerSocket.IsClosed && !innerSocket.IsInvalid)
+            {
+                // We only need to call this for true blocking operations when there is chance they are stuck in OS system call.
+                Interop.Sys.Shutdown(innerSocket, SocketShutdown.Receive);
+            }
```

Comment:
Should we just always do it?  If we don't always do it, are there not race conditions that could result in our missing doing it in some circumstances we should have?

Also, what about on Windows?  We've talked about doing a larger work item of ensuring that we don't have any sync calls pending when we close the socket, and the approach we've talked about there included calling Shutdown (plus doing some other things)... should we just do the Shutdown call now everywhere, and then augment it later?

Reply:
As far as I know, this is only needed for sockets on Unix in blocking mode where IO calls can be stuck in kernel. This change fixes three reported issues - different variants of this. 
I can open separate issue for the windows cleanup we talked about.

Reply:
> this is only needed for sockets on Unix in blocking mode where IO calls can be stuck in kernel

Sure.  My question is, is there any way some set of concurrent calls could be made with this code such that we should call Shutdown to unblock something but don't?  I'm asking whether a race condition could result where we end up not calling Shutdown but we should have to unblock something stuck.

Reply:
I think there is race condition with closing and making socket non-blocking. 
I tried to construct scenario to catch that and it is tricky as simply setting socket.Blocking = false does not do much. However that does not set  _underlyingHandleNonBlocking either. 

When socket is truly made non-blocking, the pending blocking IO wakes up with EAGAIN. 
So even if we hit the race condition we should not end-up hanging as far as I can tell.